docs/ds20.md

```diff
@@ -0,0 +1,111 @@
+We do need to use the `BOS` *“Binary Object Store”* descriptor, which is only supported for devices using USB 2.1 and newer, although that’s probably the vast majority of devices now.
+Usually the BOS descriptor is used for Wireless USB details, USB 2.0 extensions, SuperSpeed USB connection details and a Container ID.
+
+Usefully, the BOS descriptor does give us is the ability to use the platform capability descriptor, which is `bDevCapabilityType=0x05`. For UUID `D8DD60DF-4589-4CC7-9CD2-659D9E648A9F`, this is identified as a structured blob of data Windows usually uses to put workarounds like the suspend mode of the device and that kind of thing.
+
+We could certainly create a new `bDevCapabilityType` which would allow us to store a binary blob (e.g. `Plugin=foobarbaz\nFlags=QuirkValueHere\n`) but that would be out-of-specification and difficult for vendors to implement.
+
+## Proposal
+
+Devices create a new platform device capability BOS descriptor with `bDevCapabilityType=0x05` and a fwupd-specific UUID, specifically `010aec63-f574-52cd-9dda-2852550d94f0`.
+Using this custom UUID will ensure that Microsoft Windows does not try to parse the capability descriptor as a *Microsoft OS 2.0 descriptor set*.
+
+## Implementation
+
+Create a BOS DS20 descriptor such as:
+
+    1C 10 05 00 63 ec 0a 01 74 f5 cd 52 9d da 28 52 55 0d 94 f0 05 08 01 00 20 00 2a 00
+    |  |  |  |  \---------------------------------------------/ \--------/  \---/ |  |
+    |  |  |  |                ^-PlatformCapabilityUUID            |   wLength-^   |  |
+    |  |  |  ^-bReserved                                 dwVersion^   bVendorCode-^  |
+    |  |  ^-bDevCapability                                                bAltEnumCmd^
+    |  ^-bDescriptorType
+    ^-bLength
+
+The `dwVersion` encoded here is fwupd `1.8.5`, which is also the first version that supports BOS DS20 descriptors.
+A `dwVersion` of 0x0 will match any fwupd version, but should probably not be used unless required.
+
+The BOS descriptors are sorted by the requester and can appear in any order.
+The descriptor with the highest `dwVersion` that is not newer than the current running daemon version is used.
+This allows devices to set different quirks depending on the fwupd version, although in practice this should not be required.
+The `bAltEnumCmd` parameter **must** be zero.
+The `PlatformCapabilityUUID` **must** be `010aec63-f574-52cd-9dda-2852550d94f0`.
+
+Then allow the device to reply to a USB control request with the following parameters:
+
+    transfer-direction: device-to-host
+    request-type: vendor
+    recipient: device
+    bRequest: {value of bVendorCode in the BOS descriptor)
+    wValue: 0x00
+    wIndex: 0x07
+    wLength: {value of wLength in the BOS descriptor)
+
+The device should return something like:
+
+    50 6c 75 67 69 6e 3d 64 66 75 0a 49 63 6f 6e 3d 63 6f 6d 70 75 74 65 72 0a 00 00 00 00 00 00 00
+
+...which is the UTF-8 quirk data, e.g.
+
+    Plugin=dfu
+    Icon=computer
```

Comment:
The spec is missing information about how someone would "come up with" the fields to put into their descriptor.  Like a development flow.  I would think everyone should "start out" with a flat quirk file while under development and when they prove it's all working, move the quirk content into the descriptor.  Which begs a `contrib/` script that turns a quirk file into this descriptor format!

Reply:
How about `Vendors using this API should use an existing fwupd quirk file to validate the fields needed in the DS20 control transfer buffer`. I'm not 100% sure about a ./contrib/generate-descriptor.py as it's going to need almost everything specified -- and in some firmware you just want `00 63 ec 0a` i.e. not the first three bytes as the descriptor is set up in a common way.

Reply:
Even putting it into UTF-8 can be messed up if someone tries to do it by hand.  I think it's also a good litmus test to have something in fwupd source that is a unit test like this:
1. Do something needing a quirk file.
2. Use helper script that converts into a device replay buffer
3. Remove the quirk file
4. Use the device replay buffer to interact with it
5. verify it works
6. Modify the replay buffer to break it in all the common ways.

The common ways that come to mind for me from this discussion:

- Lower version than ours
- Pass in a bad GUID
- Pass in invalid quirk data

Reply:
> Use the device replay buffer to interact with it

I think we're doing that now in the new self tests.

> Lower version than ours

You mean lower fwupd version?

>  Pass in a bad GUID

You mean like a GUID specified like `Guid = FOOBARBAZ`?

>  Pass in invalid quirk data

Got any ideas on what `invalid` might be?

Reply:
> You mean lower fwupd version?

Yeah

> You mean like a GUID specified like Guid = FOOBARBAZ?

I mean like not the fwupd GUID.  If there is only one GUID used and we don't find it in the list but expected to see it throw an error up.

> Got any ideas on what invalid might be?

A plugin name that doesn't exist is probably the easiest to model.

Reply:
Done.